design/backup-performance-improvements.md

```diff
@@ -0,0 +1,211 @@
+- Process multiple backups in parallel: this is a future feature, although certain prerequisites for this enhancement are included in this proposal.
+
+## High-Level Design
+
+### Phase 1: ItemBlock processing
+- A new BIA method, `GetAdditionalItems`, will be needed for pre-processing ItemBlocks (this will require a new BIAv3 API).
+- When processing the list of items returned from the item collector, instead of simply calling `BackupItem` on each in turn, we will use the `GetAdditionalItems` BIAv3 API call to determine other items to include with the current item in an ItemBlock. Repeat recursively on each item returned.
+- Don't include an item in more than one ItemBlock -- if the next item from the item collector is already in a block, skip it.
+- Once ItemBlock is determined, call new func `BackupItemBlock` instead of `BackupItem`.
+- New func `BackupItemBlock` will call pre hooks for any pods in the block, then back up the items in the block (`BackupItem` will no longer run hooks directly), then call post hooks for any pods in the block.
+
+### Phase 2: Process ItemBlocks for a single backup in multiple threads
+- Concurrent `BackupItemBlock` operations will be executed by worker threads invoked by the backup controller, which will communicate with the backup controller operation via a shared channel.
+- The ItemBlock processing loop implemented in Phase 1 will be modified to send each newly-created ItemBlock to the shared channel rather than calling `BackupItemBlock` inline.
+- Users will be able to configure the number of workers available for concurrent `BackupItemBlock` operations.
+- Access to the BackedUpItems map must be synchronized
+
+## Detailed Design
+
+### Phase 1: ItemBlock processing
+
+#### BackupItemAction plugin changes
+
+In order for Velero to identify groups of items to back up together in an ItemBlock, we need a way to identify items which need to be backed up along with the current item. While the current `Execute` BackupItemAction method does return a list of additional items which are required by the current item, we need to know this *before* we start the item backup. To support this, we need a new API method, `GetAdditionalItems` which Velero will call on each item as it processes it for an ItemBlock. The expectation is that this method will return the same items as currently returned as additional items by the current `Execute` method, with the exception that items which are not created until calling `Execute` should not be returned here, as they don't exist yet.
+
+#### Proto changes (compiled into golang by protoc)
+
+The BackupItemAction service gets one new rpc method:
+```
+service BackupItemAction {
+    rpc GetAdditionalItems(BackupItemActionGetAdditionalItemsRequest) returns (BackupItemActionGetAdditionalItemsResponse);
+}
+```
+
+To support this new rpc method, we define new request/response message types:
```

Comment:
Given now the BIA will be called more frequently for these new calls.
For the core plugins such as Pod/PVC, can we make them simple function calls instead of being RPC calls?
We discussed this earlier in context of Resource Modifiers, that unfortunately all plugins are RPC today and this could have performance implication when called at large scale.

Reply:
@anshulahuja98 So "more frequently" means "one more call per BIA per item" in this case. That change can probably be made as a separate design/implementation, though, since it would be completely isolated from any of the design here. It won't affect the API changes or the workflow modificatons.

Reply:
@anshulahuja98 "We discussed this earlier in context of Resource Modifiers" -- is there a github issue for this? It's probably worth someone doing some benchmarking with current code, although we would potentially hit the issue sooner with the new API call. This additional RPC load could be simulated by adding another plugin or two that matches every item.

Reply:
I added this to the non-goals section. If scale testing indicates a need for it, that refactoring should be handled in a separate design and implementation, since there really shouldn't be any interaction between those changes and the ones described in this design.

Reply:
There is no current issue tracking it.
I understand keeping the 2 designs separate is the right choice for now.
But just wanted to bump up this concern. 

CC: @reasonerjt if you have any more thoughts on this.